airflow/models/dagrun.py

```diff
@@ -945,7 +945,7 @@ def _check_for_removed_or_restored_tasks(
 
     def _get_task_creator(
         self, created_counts: Dict[str, int], ti_mutation_hook: Callable, hook_is_noop: bool
-    ) -> Callable:
+    ) -> Callable[[Operator, Tuple[int, ...]], Union[Iterator[Dict[str, Any]], Iterator[TI]]]:
```

Comment:
It’s probably a good idea to make this an overload

```python
@overload
def _get_task_creator(
    self,
    created_counts: Dict[str, int],
    ti_mutation_hook: Callable,
    hook_is_noop: Literal[True],
) -> Callable[[Operator, Tuple[int, ...]], Iterator[Dict[str, Any]]]:
    ...

@overload
def _get_task_creator(
    self,
    created_counts: Dict[str, int],
    ti_mutation_hook: Callable,
    hook_is_noop: Literal[False],
) -> Callable[[Operator, Tuple[int, ...]], Iterator[TI]]:
    ...

def _get_task_creator(
    self,
    created_counts: Dict[str, int],
    ti_mutation_hook: Callable,
    hook_is_noop: bool,
) -> Callable[[Operator, Tuple[int, ...]], Union[Iterator[Dict[str, Any]], Iterator[TI]]]:
    # Actual implementation...

```

Reply:
Nice one! I never really got around to using `overloads`, but this is a very nice use case for it indeed.
Will change it :)